src/olympia/versions/models.py

```diff
@@ -1423,6 +1438,90 @@ def __str__(self):
+
+        # Recommended/line for Android are allowed to set compatibility in the
+        # limited range.
+        if (
+            addon.promoted
+            and addon.promoted.group in (RECOMMENDED, LINE)
+            and amo.ANDROID in addon.promoted.approved_applications
+        ):
+            return False
+
+        is_min_in_limited_range = self.min.version_int >= version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[0]
+        ) and self.min.version_int < version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[1]
+        )
+        is_max_in_limited_range = self.max.version_int >= version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[0]
+        ) and self.max.version_int < version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[1]
+        )
+        is_range_bigger_than_limited_range = self.min.version_int < version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[0]
+        ) and self.max.version_int >= version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[1]
+        )
+        return (
+            is_min_in_limited_range
+            or is_max_in_limited_range
+            or is_range_bigger_than_limited_range
+        )
+
+    def save(self, *args, **kwargs):
+        if (
+            self.application == amo.ANDROID.id
+            and self.version.addon.type == amo.ADDON_EXTENSION
+        ):
+            # Regardless of whether the compat information came from the
+            # manifest, on Android we currently only allow non-promoted/line
+            # extensions to be compatible with Fennec or Fenix that has general
+            # availability, so we might need to override the compat data before
+            # saving it.
+            if self.version_range_contains_forbidden_compatibility():
+                # Arbitrarily pick a range that makes sense and is valid.
+                self.min = AppVersion.objects.filter(
+                    application=amo.ANDROID.id,
+                    version=amo.MIN_VERSION_FENIX_GENERAL_AVAILABILITY,
+                ).first()
+                self.max = AppVersion.objects.filter(
+                    application=amo.ANDROID.id,
+                    version=amo.DEFAULT_WEBEXT_MAX_VERSION,
+                ).first()
+
+            # In addition we need to set/clear strict compatibility.
+            # If the range max is below the first Fenix version, then we need
+            # to set strict_compatibility to True on the File to prevent that
+            # version from being installed in Fenix. If it's not, then we need
+            # to clear it if it was set.
+            if self.max.version_int < version_int(amo.MIN_VERSION_FENIX):
+                strict_compatibility = True
+            else:
+                strict_compatibility = False
+            if self.version.file.strict_compatibility != strict_compatibility:
+                self.version.file.update(strict_compatibility=strict_compatibility)
```

Comment:
Just realized a couple issues with that I need to address:
- If someone chooses to be compatible with Fennec and gets `strict_compatibility` set to `True`, it would also apply to potential users on Desktop. So if their `max` for `FIREFOX` is set to something other than `*`, we'd obey it... I think that's an edge case we're fine with, but needs to be documented.
- I forgot to add support for deleting Firefox for Android compatibility entirely, which should drop the `strict_compatibility` if it's an extension.